File: C++/dijkstra_prime.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#define ll long long

using namespace std;

bool isPrime(ll n) {
    std::vector<ll> primes = {2};
    std::priority_queue<std::pair<ll, ll>, std::vector<std::pair<ll, ll>>, std::greater<std::pair<ll, ll>>> pool;
    pool.push({4, 2});
    
    for (ll i = 3; i <= n; ++i) {
        while (!pool.empty() && pool.top().first < i) {
            auto [multiple, prime] = pool.top();
            pool.pop();
            pool.push({multiple + prime, prime});
        }
        if (!pool.empty() && pool.top().first == i) {
            auto [multiple, prime] = pool.top();
            pool.pop();
            pool.push({multiple + prime, prime});
        } else {
            primes.push_back(i);
            pool.push({i * i, i});
        }
    }
    return find(primes.begin(), primes.end(), n) != primes.end();
}

int main() {
    std::cout << isPrime(1000000) << std::endl;
    return 0;
}

File: Prime tests/dijkstra_prime.py
import heapq

# Adapted from b001

def isPrime(n:int):
    primes = [2]
    pool = [(4, 2)]
    heapq.heapify(pool)
    for i in range(3, n+1):
        while pool[0][0] < i:
            multiple, prime = heapq.heappop(pool)
            heapq.heappush(pool, (multiple + prime, prime))
        if pool[0][0] == i:
            multiple, prime = heapq.heappop(pool)
            heapq.heappush(pool, (multiple + prime, prime))
        else:
            primes.append(i)
            heapq.heappush(pool, (i*i, i))
    return n in primes

print(isPrime(1000000))